datalad/metadata/search.py

```diff
@@ -517,8 +600,32 @@ def __call__(query=None,
         # where does the bunny have the eggs?
         index_dir = opj(ds.path, get_git_dir(ds.path), SEARCH_INDEX_DOTGITDIR)
 
+        if mode == 'default':
+            get_schema_fx = _get_schema_homoblob
+            meta2doc_fx = _meta2homoblob_dict
+            get_parser = _get_parser_homoblob
+            documenttype = ds.config.obtain(
+                'datalad.metadata.searchindex-default-documenttype',
+                default='datasets')
+        elif mode == 'autofield':
+            get_schema_fx = _get_schema_autofield
+            meta2doc_fx = _meta2autofield_dict
+            get_parser = _get_parser_autofield
+            documenttype = ds.config.obtain(
+                'datalad.metadata.searchindex-autofield-documenttype',
+                default='all')
+        else:
+            raise ValueError(
+                'unknown search mode "{}"'.format(mode))
+
         idx_obj = _get_search_index(
-            index_dir, ds, force_reindex)
+            index_dir,
+            mode,
+            ds,
+            force_reindex,
+            get_schema_fx,
+            meta2doc_fx,
+            documenttype)
```

Comment:
1c: Feels like it could have been earlier to navigate if above two types of searches were done via two classes and this generic one in a common method

Reply:
I didn't get it. What is earlier to navigate and for whom?

Reply:
Just navigate the code. Now it is a collection of functions grouped by the common prefix, which are passed around. Just felt that it would be easier to follow and use if it was classes

Reply:
I agree that this could be optimized. But we should first figure out whether we want this at all. From my perspective this is search v4 and the last three didn't make it into a release. Not the time for optimization, it seems.